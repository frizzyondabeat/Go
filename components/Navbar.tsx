import Logo from '@/public/images/logo.png';
import Profile from '@/public/images/profile.png';
import { Bell, ChevronDown, CircleHelp, Search, Settings, Wallet } from 'lucide-react';
import Image from 'next/image';
import { Input } from './ui/input';
import { cn } from '@/lib/utils';

const NavbarButtonOptions = [
    {
        name: 'Notifications',
        icon: Bell,
    },
    {
        name: 'Wallet',
        icon: Wallet,
    },
    {
        name: 'Inquiries',
        icon: CircleHelp,
    },
    {
        name: 'Settings',
        icon: Settings,
    },
];

export const Navbar = () => {
    return (
        <header className="flex h-[65px] w-full items-center justify-between border-b bg-white px-8 py-2 shadow-sm">
            <div className="flex space-x-4">
                <Image src={Logo} alt="logo" className="size-9" quality={100} />
                <div className="relative hidden h-full items-center justify-center md:flex">
                    <Input
                        placeholder="Search here..."
                        className="w-96 bg-slate-50 px-10 dark:bg-slate-50/5"
                    />
                    <Search size={18} className="absolute left-3 text-slate-500" />
                </div>
            </div>
            <div className="flex gap-5">
                {NavbarButtonOptions.map(({ name, icon: Icon }, index) => (
                    <div
                        key={`${name}-${index}`}
                        className="group flex cursor-pointer flex-col items-center justify-center gap-1"
                    >
                        <Icon
                            className={cn(
                                'size-4 transition-all duration-300 ease-in-out group-hover:text-blue-500',
                                {
                                    'group-hover:animate-wiggle': name !== 'Settings',
                                    'group-hover:animate-spin-slow': name === 'Settings',
                                }
                            )}
                        />
                        <span className="text-xs group-hover:text-blue-500">{name}</span>
                    </div>
                ))}
                <div className="flex items-center gap-1">
                    <Image src={Profile} alt="profile" className="size-9" quality={100} />
                    <ChevronDown size={18} className="text-slate-500" />
                </div>
            </div>
        </header>
    );
};
