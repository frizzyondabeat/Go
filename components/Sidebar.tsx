'use client';

import { cn } from '@/lib/utils';
import React, { FC, PropsWithChildren } from 'react';
import { Button, buttonVariants } from './ui/button';
import { ArrowLeftToLine } from 'lucide-react';
import { usePathname } from 'next/navigation';
import { SidebarItemProps } from '@/types';
import { motion } from 'framer-motion';
import Link from 'next/link';

export const Sidebar: FC<PropsWithChildren> = ({ children }) => {
    return (
        <aside className="hidden h-full w-[20%] rounded bg-white sm:flex">
            <nav className="flex h-full w-full flex-col p-3 shadow-lg dark:text-white">
                <span className="text-xs font-semibold">Settings</span>

                <ul className={cn('flex-1 space-y-2 px-4')}>{children}</ul>

                <div className="flex w-full items-center justify-center">
                    <Button variant={'outline'}>
                        <ArrowLeftToLine className="mr-2 size-4" />
                        <span>Back to Dashboard</span>
                    </Button>
                </div>
            </nav>
        </aside>
    );
};

export const SidebarItem: FC<SidebarItemProps> = ({ icon: Icon, href, label }) => {
    const pathname = usePathname();

    const isActive =
        (pathname === '/' && href === '/') || pathname === href || pathname?.startsWith(`${href}/`);

    const MotionLink = motion(Link);

    const isSettings = label.includes('Settings');

    return (
        <li className="group relative flex items-center justify-between">
            <MotionLink
                whileHover={{ scale: !isActive ? 1.05 : 1 }}
                transition={{ duration: 0.3 }}
                href={href}
                className={cn(
                    buttonVariants({ variant: 'ghost' }),
                    'hover:bg-primary-100/15 relative w-full justify-start rounded-sm p-0 transition-colors duration-300'
                )}
            >
                {isActive && (
                    <motion.div
                        layoutId="active-tab"
                        className="absolute inset-0 rounded-r-md bg-gradient-to-r from-transparent to-blue-500/10 dark:to-blue-200/30"
                    />
                )}
                <Icon
                    className={cn(
                        'size-5 stroke-[1.5] transition-all duration-200 group-hover:delay-100',
                        {
                            '-rotate-6 stroke-2 text-blue-700 dark:text-white': isActive,
                            'group-hover:animate-spin-slow': isSettings && !isActive,
                            'group-hover:animate-wiggle': !isSettings && !isActive,
                        }
                    )}
                />
                <span
                    className={cn('ml-5 overflow-hidden text-xs font-normal transition-all', {
                        'font-medium text-blue-700 dark:text-white': isActive,
                    })}
                >
                    {label}
                </span>
                {isActive && (
                    <div className="absolute right-0 h-10 w-1 rounded-r-full bg-blue-700 dark:bg-blue-200" />
                )}
            </MotionLink>
        </li>
    );
};
