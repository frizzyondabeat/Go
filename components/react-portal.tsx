import { FC, PropsWithChildren, useLayoutEffect, useState } from 'react';
import { createPortal } from 'react-dom';
import { AnimatePresence } from 'framer-motion';

type ReactPortalProps = PropsWithChildren & {
    wrapperId: string;
};

const createWrapperAndAppendToBody = (wrapperId: string) => {
    if (!document) return null;
    const wrapperElement = document.createElement('div');
    wrapperElement.setAttribute('id', wrapperId);
    document.body.appendChild(wrapperElement);
    return wrapperElement;
};

/**
 * **Description:**
 *
 * ReactPortal component is used to create a portal for the children to be rendered in a different part of the DOM.
 *
 * **Usage:**
 *
 * ```tsx
 *
 * <ReactPortal wrapperId="portal-wrapper">
 *
 *      <div>Portal Content</div>
 *
 * </ReactPortal>
 *
 * ```
 * @param {any} children
 * @param {any} wrapperId
 * @returns {any}
 */
export const ReactPortal: FC<ReactPortalProps> = ({ children, wrapperId }) => {
    const [wrapperElement, setWrapperElement] = useState<HTMLElement>();

    useLayoutEffect(() => {
        let element = document.getElementById(wrapperId);
        let systemCreated = false;
        if (!element) {
            systemCreated = true;
            element = createWrapperAndAppendToBody(wrapperId);
        }
        setWrapperElement(element!);

        return () => {
            if (systemCreated && element?.parentNode) {
                element?.parentNode.removeChild(element);
            }
        };
    }, [wrapperId]);

    if (!wrapperElement) return null;

    return createPortal(<AnimatePresence mode="wait">{children}</AnimatePresence>, wrapperElement);
};
