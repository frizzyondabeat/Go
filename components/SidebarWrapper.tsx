'use client';

import { sidebarRoutes } from '@/utils/helpers';
import { Sidebar, SidebarItem } from './Sidebar';

export const SidebarWrapper = () => {
    return (
        <Sidebar>
            {sidebarRoutes.map((route, index) => (
                <SidebarItem key={index} {...route} />
            ))}
        </Sidebar>
    );
};
