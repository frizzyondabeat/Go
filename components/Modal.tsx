'use client';
import React, { FC, PropsWithChildren, useRef } from 'react';

import { motion, AnimatePresence } from 'framer-motion';
import { useDisableScroll, useOnClickOutside } from '@/hooks';
import { ReactPortal } from './react-portal';
import { cn } from '@/lib/utils';
import { Button } from './ui/button';
import { X } from 'lucide-react';

type ModalProps = PropsWithChildren & {
    isOpen: boolean;
    onClose: () => void;
};

export const ModalContext = React.createContext<ModalProps>({
    isOpen: false,
    onClose: () => {},
});

const Modal: FC<ModalProps> & {
    Container: FC<ModalContainerProps>;
    Content: FC<ModalContentProps>;
    Header: FC<ModalHeaderProps>;
    Body: FC<ModalBodyProps>;
    Footer: FC<ModalFooterProps>;
} = ({ children, isOpen, onClose }) => {
    return (
        isOpen && (
            <ReactPortal wrapperId="modal">
                <ModalContext.Provider value={{ isOpen, onClose }}>
                    {children}
                </ModalContext.Provider>
            </ReactPortal>
        )
    );
};

export const useModal = () => {
    if (!React.useContext(ModalContext)) {
        throw new Error('useModal must be used within a ModalProvider');
    }
    return React.useContext(ModalContext);
};

export type ModalContainerProps = PropsWithChildren & {
    className?: string;
};

const ModalContainer: FC<ModalContainerProps> = ({ children, className }) => {
    return (
        <motion.div
            key="modal-container"
            initial={{ opacity: 0 }}
            animate={{ opacity: 1 }}
            exit={{ opacity: 0 }}
            transition={{ duration: 0.3 }}
            className={cn(
                'fixed inset-0 z-50 flex items-center justify-center bg-black bg-opacity-50 backdrop-blur-[1px]',
                className
            )}
        >
            {children}
        </motion.div>
    );
};

export type ModalContentProps = PropsWithChildren & {
    className?: string;
    animateDirection?: 'from-bottom' | 'from-right';
};

const ModalContent: FC<ModalContentProps> = ({
    children,
    className,
    animateDirection = 'from-bottom',
}) => {
    const ref = useRef<HTMLDivElement>(null);
    const { onClose: toggleModal } = useModal();
    useOnClickOutside(ref, toggleModal);
    useDisableScroll();

    return (
        <motion.div
            ref={ref}
            key="modal-content"
            initial={{
                opacity: 0,
                y: animateDirection === 'from-bottom' ? 50 : 0,
                x: animateDirection === 'from-right' ? 50 : 0,
            }}
            animate={{ opacity: 1, y: 0, x: 0 }}
            exit={{
                opacity: 0,
                y: animateDirection === 'from-bottom' ? 50 : 0,
                x: animateDirection === 'from-right' ? 50 : 0,
            }}
            transition={{ duration: 0.3 }}
            className={cn(
                'max-h-screen w-full max-w-lg rounded-xl bg-white p-4 shadow-md',
                className
            )}
        >
            {children}
        </motion.div>
    );
};

export type ModalHeaderProps = PropsWithChildren & {
    className?: string;
    type?: 'HEADLESS' | 'PRIMARY' | 'SECONDARY' | 'CUSTOM' | 'EMPTY';
    title?: string;
    subtitle?: string;
    closeIconClass?: string;
    closeIconBgClass?: string;
};

const ModalHeader: FC<ModalHeaderProps> = ({
    children,
    className,
    type,
    title,
    closeIconClass,
    closeIconBgClass,
    subtitle,
}) => {
    const { onClose: toggleModal } = useModal();
    const renderHeadless = () => (
        <div className={cn('flex w-full justify-end px-5', className)}>
            <Button
                onClick={toggleModal}
                variant="destructive"
                size={'icon'}
                className={cn('p-1', closeIconBgClass)}
            >
                <X className={cn('size-4 stroke-1', closeIconClass)} />
            </Button>
        </div>
    );

    const renderPrimary = () => (
        <div
            className={cn(
                'flex w-full items-center justify-between border-b px-4 pb-4 pt-0',
                className
            )}
        >
            <div className="flex flex-col leading-3">
                <span className="text-base font-medium">{title}</span>
                <h3 className="text-sm font-light text-gray-400">{subtitle}</h3>
            </div>
            <Button
                onClick={toggleModal}
                variant="destructive"
                size={'icon'}
                className={cn('p-1', closeIconBgClass)}
            >
                <X className={cn('size-4 stroke-1', closeIconClass)} />
            </Button>
        </div>
    );

    const renderCustom = () => (
        <div className={cn('relative flex w-full items-center justify-center p-4', className)}>
            {children}
            <Button
                onClick={toggleModal}
                variant="destructive"
                size={'icon'}
                className={cn('absolute right-2 top-2 p-1', closeIconBgClass)}
            >
                <X className={cn('size-4 stroke-1', closeIconClass)} />
            </Button>
        </div>
    );

    const renderEmpty = () => <></>;

    const renderMap: Record<string, () => JSX.Element> = {
        HEADLESS: renderHeadless,
        PRIMARY: renderPrimary,
        CUSTOM: renderCustom,
        EMPTY: renderEmpty,
    };

    const render = renderMap[type!] || renderEmpty;

    return <>{render()}</>;
};

export type ModalBodyProps = PropsWithChildren & {
    className?: string;
};

const ModalBody: FC<ModalBodyProps> = ({ children, className }) => {
    return <div className={cn('mt-4', className)}>{children}</div>;
};

export type ModalFooterProps = PropsWithChildren & {
    className?: string;
};

const ModalFooter: FC<ModalFooterProps> = ({ children, className }) => {
    return <div className={cn('mt-4 flex w-full border-t px-4', className)}>{children}</div>;
};

Modal.Container = ModalContainer;
Modal.Content = ModalContent;
Modal.Header = ModalHeader;
Modal.Body = ModalBody;
Modal.Footer = ModalFooter;

export { Modal };
