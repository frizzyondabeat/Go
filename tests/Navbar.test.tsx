import { describe, expect, it } from 'vitest';
import { Navbar } from '@/components/Navbar';
import { fireEvent, render, screen } from '@testing-library/react';

declare const global: {
    NavbarButtonOptions: any[];
};

describe('unit tests', () => {
    it('should render the Navbar component', () => {
        const { container } = render(<Navbar />);
        expect(container).toMatchSnapshot();
    });

    it('should display alt text when profile image fails to load', () => {
        const { container } = render(<Navbar />);
        const image = container.querySelector('img[alt="logo"]') as HTMLImageElement;
        fireEvent.error(image);
        expect(image.alt).toBe('logo');
    });

    it('should render correctly with an empty NavbarButtonOptions array', () => {
        const originalNavbarButtonOptions = global.NavbarButtonOptions;
        global.NavbarButtonOptions = [];
        const { container } = render(<Navbar />);
        expect(container).toBeDefined();
        global.NavbarButtonOptions = originalNavbarButtonOptions;
    });

    it('should have appropriate aria-labels for accessibility', () => {
        render(<Navbar />);
        const inputs = screen.getAllByLabelText('Search');

        inputs.forEach(input => {
            expect(input).toBeDefined();
        });
    });
});
