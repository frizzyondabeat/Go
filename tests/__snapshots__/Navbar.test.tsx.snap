// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`unit tests > should render the Navbar component 1`] = `
<div>
  <header
    class="flex h-[65px] w-full items-center justify-between border-b bg-white px-8 py-2 shadow-sm"
  >
    <div
      class="flex space-x-4"
    >
      <img
        alt="logo"
        class="size-9"
        data-nimg="1"
        decoding="async"
        height="49"
        loading="lazy"
        src="/_next/image?url=%2Fpublic%2Fimages%2Flogo.png&w=128&q=100"
        srcset="/_next/image?url=%2Fpublic%2Fimages%2Flogo.png&w=64&q=100 1x, /_next/image?url=%2Fpublic%2Fimages%2Flogo.png&w=128&q=100 2x"
        style="color: transparent;"
        width="50"
      />
      <div
        class="relative hidden h-full items-center justify-center md:flex"
      >
        <input
          aria-label="Search"
          class="flex h-10 rounded-md border border-input py-2 text-sm ring-offset-background file:border-0 file:bg-transparent file:text-sm file:font-medium placeholder:text-muted-foreground focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:cursor-not-allowed disabled:opacity-50 w-96 bg-slate-50 px-10 dark:bg-slate-50/5"
          name="search"
          placeholder="Search here..."
        />
        <svg
          class="lucide lucide-search absolute left-3 text-slate-500"
          fill="none"
          height="18"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="18"
          xmlns="http://www.w3.org/2000/svg"
        >
          <circle
            cx="11"
            cy="11"
            r="8"
          />
          <path
            d="m21 21-4.3-4.3"
          />
        </svg>
      </div>
    </div>
    <div
      class="flex gap-5"
    >
      <div
        class="group flex cursor-pointer flex-col items-center justify-center gap-1"
      >
        <svg
          class="lucide lucide-bell size-4 transition-all duration-300 ease-in-out group-hover:text-blue-500 group-hover:animate-wiggle"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M6 8a6 6 0 0 1 12 0c0 7 3 9 3 9H3s3-2 3-9"
          />
          <path
            d="M10.3 21a1.94 1.94 0 0 0 3.4 0"
          />
        </svg>
        <span
          class="text-xs group-hover:text-blue-500"
        >
          Notifications
        </span>
      </div>
      <div
        class="group flex cursor-pointer flex-col items-center justify-center gap-1"
      >
        <svg
          class="lucide lucide-wallet size-4 transition-all duration-300 ease-in-out group-hover:text-blue-500 group-hover:animate-wiggle"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M19 7V4a1 1 0 0 0-1-1H5a2 2 0 0 0 0 4h15a1 1 0 0 1 1 1v4h-3a2 2 0 0 0 0 4h3a1 1 0 0 0 1-1v-2a1 1 0 0 0-1-1"
          />
          <path
            d="M3 5v14a2 2 0 0 0 2 2h15a1 1 0 0 0 1-1v-4"
          />
        </svg>
        <span
          class="text-xs group-hover:text-blue-500"
        >
          Wallet
        </span>
      </div>
      <div
        class="group flex cursor-pointer flex-col items-center justify-center gap-1"
      >
        <svg
          class="lucide lucide-circle-help size-4 transition-all duration-300 ease-in-out group-hover:text-blue-500 group-hover:animate-wiggle"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <circle
            cx="12"
            cy="12"
            r="10"
          />
          <path
            d="M9.09 9a3 3 0 0 1 5.83 1c0 2-3 3-3 3"
          />
          <path
            d="M12 17h.01"
          />
        </svg>
        <span
          class="text-xs group-hover:text-blue-500"
        >
          Inquiries
        </span>
      </div>
      <div
        class="group flex cursor-pointer flex-col items-center justify-center gap-1"
      >
        <svg
          class="lucide lucide-settings size-4 transition-all duration-300 ease-in-out group-hover:text-blue-500 group-hover:animate-spin-slow"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M12.22 2h-.44a2 2 0 0 0-2 2v.18a2 2 0 0 1-1 1.73l-.43.25a2 2 0 0 1-2 0l-.15-.08a2 2 0 0 0-2.73.73l-.22.38a2 2 0 0 0 .73 2.73l.15.1a2 2 0 0 1 1 1.72v.51a2 2 0 0 1-1 1.74l-.15.09a2 2 0 0 0-.73 2.73l.22.38a2 2 0 0 0 2.73.73l.15-.08a2 2 0 0 1 2 0l.43.25a2 2 0 0 1 1 1.73V20a2 2 0 0 0 2 2h.44a2 2 0 0 0 2-2v-.18a2 2 0 0 1 1-1.73l.43-.25a2 2 0 0 1 2 0l.15.08a2 2 0 0 0 2.73-.73l.22-.39a2 2 0 0 0-.73-2.73l-.15-.08a2 2 0 0 1-1-1.74v-.5a2 2 0 0 1 1-1.74l.15-.09a2 2 0 0 0 .73-2.73l-.22-.38a2 2 0 0 0-2.73-.73l-.15.08a2 2 0 0 1-2 0l-.43-.25a2 2 0 0 1-1-1.73V4a2 2 0 0 0-2-2z"
          />
          <circle
            cx="12"
            cy="12"
            r="3"
          />
        </svg>
        <span
          class="text-xs group-hover:text-blue-500"
        >
          Settings
        </span>
      </div>
      <div
        class="flex items-center gap-1"
      >
        <img
          alt="profile"
          class="size-9"
          data-nimg="1"
          decoding="async"
          height="36"
          loading="lazy"
          src="/_next/image?url=%2Fpublic%2Fimages%2Fprofile.png&w=96&q=100"
          srcset="/_next/image?url=%2Fpublic%2Fimages%2Fprofile.png&w=48&q=100 1x, /_next/image?url=%2Fpublic%2Fimages%2Fprofile.png&w=96&q=100 2x"
          style="color: transparent;"
          width="36"
        />
        <svg
          class="lucide lucide-chevron-down text-slate-500"
          fill="none"
          height="18"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="18"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="m6 9 6 6 6-6"
          />
        </svg>
      </div>
    </div>
  </header>
</div>
`;
